### ISSUE TITLE
Macro to remove setup boilerplate

### ISSUE THREAD

**Issue by JoshMcguigan** [purpliminal/rust-dotenv#98]
Thanks for writing this library, I've found it very helpful. One thing I have noticed though is that I end up with a lot of boilerplate at the top of my code.

```
    let password = dotenv::var("PASSWORD").unwrap();
    let client_id = dotenv::var("CLIENT_ID").unwrap();
    let client_secret = dotenv::var("CLIENT_SECRET").unwrap();
```

I know dotenv has a macro, but I'm not sure if it would help in this case? Have you considered a macro that could cut down on this boilerplate?

```
macro_rules! dotenv_map {
        ($name:ident) => {
            let $name =dotenv::var(stringify!($name).to_uppercase()).unwrap();
        };
    }

fn main() {
    dotenv_map!(foo);
    println!("foo: {:?}", foo);
}
```

I'd be happy to submit this as a PR if there is interest.

**Comment by kamek-pf** [comment]
I'm looking for the same thing.
I've played with the `dotenv!` macro, it checks if the environment variables you're trying to access are defined at compile time, and then it writes the values directly.

You end up with the variables you had at compile time statically built in the output binary. 
This means the only way to change the environment variables is to recompile the program with a new set of variables.
On the flip side, you get to skip a hashmap lookup, so it's a tradeoff.

I'd be interrested in something in-between what you're proposing and what the current macro does. 
Let's say `dotenv_lazy!("MY_VAR")` would :
- Check variables at compile time (like `dotenv!` does)
- Still read from the environment at run time
- Panic at run time on undefined environment variable access (like what you're proposing)

Thoughts ?

**Comment by purpliminal** [comment]
Hi! Thank you both for your suggestions.

@kamek-pf The purpose of writing the values at compile time is not for performance reasons, but for them to embed values from environment variables at build time. Think of it as environment variable-based compiler flags. I don't understand what the purpose of checking if a variable exists at compile time but reading it at run time is; it may not exist by the time the program is ran. It doesn't even have to be the same computer. I can't think of an use case for this, but I may be missing something.

@JoshMcguigan I feel like the macro you propose makes very specific assumptions (i.e. environment variable keys being uppercased and environment variable keys being representable as Rust variable names) that would be better suited as part of a configuration crate. Consider suggesting an improvement to filipegoncalves/rust-config?

**Comment by kamek-pf** [comment]
Now that I think of it my situation is probably a bit unusual. I work on a medium sized API with lots of environment variable, our team is about 10 people.
During development, we frequently need to swap some of these variables, and having to recompile because of this is a bit cumbersome.

One of the pain points is that whenever someone adds a new variable, other team members sometimes end up with errors if they are unaware of the change. So having that compile time check would be valuable to us.

I agree that it might be more confusing than helpful to implement this here though.

**Comment by purpliminal** [comment]
Right! I had to solve a similar problem at $lastjob (in Go, not Rust) and we chose to go with a simple helper function like:

```go
func mustGetenv(key string) (value string) {
  value = os.Getenv(key); if value == "" {
    log.Fatalf("Key %s not found", key) // this panics
  }
}
```

My point is, I don't see why this needs to be a macro at all. You don't care about whether your team members applied the environment variables _on compile time_; you might as well find out on runtime, provided your application blows up when the environment variable is not set.
